lib/Core/BuildEngine.cpp

```diff
@@ -931,31 +942,97 @@ class BuildEngineImpl {
         didWork = true;
       }
 
-      // If we didn't do any work, we are done.
-      if (!didWork)
-        break;
-    }
+      if (!didWork) {
+        // If there was no work to do, but we still have running tasks, then
+        // we have found a cycle. Try to resolve it and continue.
+        if (!taskInfos.empty()) {
+          if (resolveCycle(buildKey)) {
+            continue;
+          } else {
+            cancelRemainingTasks();
+            return false;
+          }
+        }
 
-    // If there was no work to do, but we still have running tasks, then
-    // we have found a cycle and are deadlocked.
-    if (!taskInfos.empty()) {
-      reportCycle(buildKey);
-      return false;
+        // We didn't do any work, and we have nothing more we can/need to do.
+        break;
+      }
     }
 
     return true;
   }
 
-  /// Report the cycle which has called the engine to be unable to make forward
+  /// Attempt to resolve a cycle which has called the engine to be unable to make forward
   /// progress.
   ///
   /// \param buildKey The key which was requested to build (the reported cycle
   /// with start with this node).
-  void reportCycle(const KeyType& buildKey) {
+  /// \returns True if the engine should try to proceed, false if the build the could not
+  /// be broken.
+  bool resolveCycle(const KeyType& buildKey) {
     // Take all available locks, to ensure we dump a consistent state.
     std::lock_guard<std::mutex> guard1(taskInfosMutex);
     std::lock_guard<std::mutex> guard2(finishedTaskInfosMutex);
 
+    std::vector<Rule*> cycleList = findCycle(buildKey);
+    assert(!cycleList.empty());
+
+    if (delegate.shouldResolveCycle(cycleList)) {
+      TracingInterval i(EngineQueueItemKind::BreakingCycle);
+
+      for (auto ruleIt = cycleList.rbegin(); ruleIt != cycleList.rend(); ruleIt++) {
```

Comment:
It would be nice to have a comment here (or somewhere) explaining the basic strategy... for example how/why we start with the node we start with.